hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/payload/ExpressionPayload.scala

```diff
@@ -164,6 +163,34 @@ class ExpressionPayload(@transient record: GenericRecord,
     }
   }
 
+  private def doRecordMerge(incomingRecord: GenericRecord,
+                            existingRecord: IndexedRecord,
+                            schema: Schema,
+                            properties: Properties): HOption[IndexedRecord] = {
+    val originalPayload = properties.getProperty(PAYLOAD_ORIGINAL_AVRO_PAYLOAD)
+    if (originalPayload.equals(classOf[OverwriteWithLatestAvroPayload].getName)) {
+      HOption.of(incomingRecord)
+    } else if (originalPayload.equals(classOf[DefaultHoodieRecordPayload].getName)) {
+      if (needUpdatingPersistedRecord(existingRecord, incomingRecord, properties)) {
+        HOption.of(incomingRecord)
+      } else {
+        HOption.of(existingRecord)
+      }
+    } else {
+      val orderingField = ConfigUtils.getOrderingField(properties)
+      if (StringUtils.isNullOrEmpty(orderingField)) {
+        HOption.of(incomingRecord)
+      } else {
+        val consistentLogicalTimestampEnabled = properties.getProperty(KeyGeneratorOptions.KEYGENERATOR_CONSISTENT_LOGICAL_TIMESTAMP_ENABLED.key,
+          KeyGeneratorOptions.KEYGENERATOR_CONSISTENT_LOGICAL_TIMESTAMP_ENABLED.defaultValue).toBoolean
+        val incomingRecordPayload = HoodieRecordUtils.loadPayload(originalPayload, incomingRecord,
+          HoodieAvroUtils.getNestedFieldVal(incomingRecord, orderingField, true, consistentLogicalTimestampEnabled)
+            .asInstanceOf[Comparable[_]]).asInstanceOf[HoodieRecordPayload[_ <: HoodieRecordPayload[_]]]
+        incomingRecordPayload.combineAndGetUpdateValue(existingRecord, schema, properties)
```

Comment:
Could we just cache `incomingRecordPayload` for all payload classes and invoke `incomingRecordPayload.combineAndGetUpdateValue` for `OverwriteWithLatestAvroPayload` and `DefaultHoodieRecordPayload` too?

Reply:
The payload takes in the record as a param so I'm not sure we can do any caching. It is also still functionally correct if we get rid of the special handling for overwrite and default payloads and load the payload for everything. This could be a case of over optimization but I would think getting rid of reflection that happens for every record would be non negligible. So that's why I did it